plat/kvm/arm/bpt64.S

```diff
@@ -0,0 +1,183 @@
+ *
+ * @param paddr physical address (not used)
+ * @param pages number of PTEs
+ */
+.macro pte_zero paddr=0, pages
+	.fill \pages, 0x8, 0
+.endm
+
+/* ----------------------- KVM MEMORY MAP (QEMU virt) ---------------------
+ *
+ * 0x0000000000000000 - 0x0000000007ffffff	Hole:          0    - 128MiB
+ * 0x0000000008000000 - 0x000000003fffffff	Devices:     128MiB -   1GiB
+ * 0x0000000040000000 - 0x0000007fffffffff	Kernel:        1GiB - 512GiB
+ * 0x0000008000000000 - 0x000000ffffffffff	PCIe hi mem: 512GiB -   1TiB
+ *
+ * Notice: The page tables below use the Unikraft indexing convention (x86).
+ */
+.section .data
+
+.global arm64_bpt_l3_pt0
+
+/* L3: 0 - 2TiB (512GiB / entry)
+ *
+ * 0x0000000000000000 - 0x0000007fffffffff	Table descriptor to l2_pt0
+ * 0x0000008000000000 - 0x000000ffffffffff	Table descriptor to l2_pt1
+ * 0x0000010000000000 - 0x0000ff7fffffffff 	Unmapped
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Table descriptor to l2_pt511
+ */
+.align 12
+arm64_bpt_l3_pt0:
+	pte_tbl_desc    arm64_bpt_l2_pt0
+	pte_tbl_desc    arm64_bpt_l2_pt1
+	pte_zero        , 509
+#ifdef CONFIG_PAGING
+	pte_tbl_desc	arm64_bpt_l2_pt511
+#endif /* CONFIG_PAGING */
+
+/* L2: 0 - 512GiB (1GiB / entry)
+ *
+ * 0x0000000000000000 - 0x000000003fffffff	Table descriptor to l1_pt0
+ * 0x0000000040000000 - 0x000000007fffffff	Table descriptor to l1_pt1
+ * 0x0000000080000000 - 0x000000003fffffff	RAM       @   2GiB
+ * 0x0000000400000000 - 0x0000007fffffffff	PCIe ECAM @ 256GiB
+ */
+.align 12
+arm64_bpt_l2_pt0:
+	pte_tbl_desc    arm64_bpt_l1_pt0
+	pte_tbl_desc    arm64_bpt_l1_pt1
+	pte_fill        0x0000000080000000, 254, 2, SECT_ATTR_NORMAL
+	pte_fill        0x0000004000000000, 256, 2, SECT_ATTR_DEVICE_nGnRnE
+
+/* L2: 512GiB - 1TiB (1GiB / entry)
+ *
+ * 0x0000008000000000 - 0x000000ffffffffff	PCIe hi-mem @ 512GiB
+ */
+.align 12
+arm64_bpt_l2_pt1:
+	pte_fill        0x0000008000000000, 512, 2, SECT_ATTR_DEVICE_nGnRE
+
+#ifdef CONFIG_PAGING
+/* L2: 255.5 TiB - 256TiB (1GiB / entry)
+ *
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Direct-mapped
+ */
+.align 12
+arm64_bpt_l2_pt511:
+	pte_fill	0x0000000000000000, 512, 2, SECT_ATTR_NORMAL
+#endif /* CONFIG_PAGING */
+
+/* L1: 0 - 1GiB (2MiB / entry)
+ *
+ * 0x0000000000000000 - 0x0000000007ffffff	Hole:   @ 0
+ * 0x0000000008000000 - 0x000000003fffffff	Devices @ 128MiB
+ */
+.align 12
+arm64_bpt_l1_pt0:
+	pte_zero        0x0000000000000000, 64
+	pte_fill        0x0000000008000000, 448, 1, SECT_ATTR_DEVICE_nGnRE
+
+/* L1: 1GiB - 2GiB (2MiB per entry)
+ *
+ * 0x0000000000000000 - 0x00000000001fffff	Table descriptor to l0_pt0
+ * 0x0000000000200000 - 0x0000000000000000	RAM	@ 1.2GiB
+ */
+.align 12
+arm64_bpt_l1_pt1:
+	pte_tbl_desc  arm64_bpt_l0_pt0
+	pte_fill      0x0000000040200000, 511, 1, SECT_ATTR_NORMAL
+
+/* L0: 1GiB - 1.2GiB (4KiB / entry)
+ *
+ * 0x0000000040000000 - 0x00000000401fffff	Kernel	@ 1GiB
+ */
+.align 12
+arm64_bpt_l0_pt0:
+	pte_fill        0x0000000040000000, 512, 0, SECT_ATTR_NORMAL_RWX
+
```

Comment:
Open issue as mentioned by @mariasfiraiala @eduardvintila
Executable code for sqlite reaches into above `0x40200000`. Open for discussion

Reply:
ACK, as discussed. Since the kernel image starts at 1GiB, will update the remaining memory of `arm64_bpt_l1_pt1` to `NORMAL_RWX`, ie to the 2GiB boundary.